### ISSUE TITLE
steps of evaluated best checkpoint

### ISSUE THREAD

**Issue by saareliad** [google-research/text-to-text-transfer-transformer#349]
Hi,
We are trying to replicate some fine-tuning the results of the paper
but don't have the resources to finetune for 2^18 steps and evaluate all checkpoints (it is often not needed).

Can you share exact number of steps you used for the best checkpoint?
Specifically, I'm interested in the larger models 3B, 11B for squad/glue.

**Comment by craffel** [comment]
Hi, for many of the tasks it was indeed necessary to fine-tune for a long time to reach the max validation accuracy. For example, for 11B on QQP, we used the checkpoint after fine-tuning for 210k steps. You may be able to save some compute by fine-tuning on the mixture, though again for some of the tasks the max validation performance happened around 200k fine-tune steps and you will sacrifice some accuracy doing multi-task training there. If it would still be very helpful for you to know the fine-tune steps, I can tabulate them - just confirming they woudl actually be useful since it's a little bit of a pain in to retrieve them all.

**Comment by saareliad** [comment]
Yes it would be extremely helpful (and interesting): we want to replicate several fine-tuning experiments with different distributed settings and hardware.
We are specifically interested in restoring sota generalization results or getting close enough with similar hyperparameters (batch size, steps and so on).

It would also help us to better choose which experiments to replicate.

We plan to opensource our code/system very soon so I hope the community would benefit as well.

**Comment by craffel** [comment]
Hi, ok. Here are fine-tune steps for all of the tasks you asked about for the 11B model:

```
Task     | CoLA | SST-2 | MRPC | STSB | QQP    | MNLIm | QNLI  | GLUE RTE | SQuAD 
FT steps | 7000 | 34000 | 7000 | 3000 | 210000 | 10000 | 20000 | 4000     | 4000 
```

Note that for WNLI we use the WSC/DPR checkpoint, which was trained for 34k steps.

I can try to get the numbers for 3B later on but they should be similar as for the 11B model. Note that your mileage may vary as far as the exact number of steps so you should probably fine-tune for longer for the sake of model selection. I'd also suggest saving checkpoints as often as you feasibly can, particularly for the very-low-resource tasks.

**Comment by saareliad** [comment]
Thanks @craffel !
To make the picture whole, I want to make sure: which batch sizes/tokens per batch did you use for these models/tasks? Did you use gradient accumulation?

For example, in the paper, you wrote you used batch size 8 for glue/superglue when fine-tuning on individual tasks because 2^11 was too high, and in the "base" experiments you used a batch size of 128.
In QA notebook for example it is a different batch size per model.

**Comment by craffel** [comment]
It is as stated in the paper.

**Comment by saareliad** [comment]
Hi @craffel, I managed to almost-restore RTE accuracy for the T5-3B model, got 90.97% after 4500 steps, best checkpoint was ~2330 steps. Paper is 91.1% for that model (I think this is 1 validation example difference).

However, it took tons of time so it is definitely not feasible for us to replicate everything on our hardware.

Are there other tasks that I did not mention which have a relatively small number of training steps/time? (e.g <=10K)

- cnn_dailymail?
- super_glue?
- wmt*?
- trivia_qa?

we start only from the released checkpoints

**Comment by adarob** [comment]
I used 10k for the CBQA tasks (https://github.com/google-research/google-research/tree/master/t5_closed_book_qa/t5_cbqa), although they mostly plateau after only a few hundred steps.

**Comment by craffel** [comment]
@saareliad if you are referring to the pre-trained multi-task checkpoints, all of them will have good performance on cnndm/squad/wmt out of the box since they were pre-trained on a multi-task mixture including those tasks. But to get the best performance, you need to fine-tune for a long time since those are all large datasets. The only small datasets are in the GLUE/SuperGLUE benchmarks.

**Comment by saareliad** [comment]
I see. I notice that in T5 the models need more finetuning than previous transformers, for example, glue RTE 4000 steps is ~103 epochs, BERT large does 2 epochs, BERT base does 3 epochs.
T5/adagrad does not use the first momentum. I tried adding momentum but it didn't reduce time-to-sota-accuracy.
I trained on squad before, I don't remember it being "long" (~20 minutes with BERT-large with our training & commodity hardware) so I considered it feasible.
BART CNN/DM is ~5 hours according to huggingface so I considered it feasible as well.

It would be extremely helpful for us if you can share a similar table to super-glue tasks (or any task if any) that are feasible to reproduce on commodity hardware.

**Comment by craffel** [comment]
> I notice that in T5 the models need more finetuning than previous transformers, for example, glue RTE 4000 steps is ~103 epochs, BERT large does 2 epochs, BERT base does 3 epochs.

You asked for the *best* checkpoint. T5's performance jumps up extremely quickly and then hovers around the best value indefinitely; it's probably mostly noise that it happened at 4k steps. I bet the same would happen with BERT if you just continued fine-tuning for longer. If you just want to get good (not "best") performance, you can fine-tune for much less time.

Here are the steps for SuperGLUE: 
```
BoolQ    CB      CoPA   MultiRC   ReCoRD   RTE      WiC     WSC
3000     1000    3000	17000     30000    2000     17000   34000
```
